Android vision document (#2946)
apple/swift-evolution: visions/android.md

While Kotlin remains the recommended language for Android development, Swift's expansion offers a compelling choice for developers seeking a modern language with strong performance and safety guarantees. Ongoing work will focus on making Swift a first-class citizen on Android, ensuring that Swift applications feel native and perform optimally on the platform.

## The Android platform

For those unfamiliar with Android's technical underpinnings, it uses a forked linux kernel optimized for mobile use, replacing the GNU libc with its own Bionic libc and dynamic linker. There is no libc++ available on-device for all apps to use, but the Android Native Development Kit (NDK) comes with a version of the LLVM libc++ that can be bundled with each app that needs it. Most apps are written in Java or Kotlin and run on the Android RunTime (ART), distributed as bytecode and either compiled to machine code when installed or Just-In-Time (JIT) when run.

> C and C++ code are compiled Ahead-Of-Time (AOT) and packaged as native shared libraries- that's how Swift on Android is also distributed- which have to be accessed through the Java Native Interface (JNI), since most Android APIs are only made available through Java.

Reviewer: I think that the distribution mechanism is not entirely tied indefinitely. While we should encourage dynamic linking, I believe that it should be possible to perform static linking. More importantly, the piece that is core to the message here is that JNI is a requirement for interop.

```suggestion
Because most Android APIs are only available through Java, applications wishing to use native code must use Java Native Interface (JNI).
```

But, this feels like an incomplete thought. Okay, so it requires JNI, but what does that mean for the vision?

Reviewer 2: I mean, it specifically says it is how it is "distributed," and as you well know, the Android shared library we distribute can either statically or dynamically link the Swift runtime and user code into it.

If you think more precise language is needed here, please suggest some.

> All four of the Android target architectures work well with Swift- 64-bit arm64-v8a and x86_64 and 32-bit armeabi-v7a and x86- and we will support [the new RISC-V riscv64 arch](https://github.com/google/android-riscv64#status) as it gets rolled out.

Reviewer 2: I don't like these changes: no real reason to break out the arches in a full list like this, and it is incorrect to say rv64 is currently supported, as there is _no_ shipping hardware yet.

I don't think we should "support" all of them with the SDK either, as x86 is barely used and they have been contemplating when to drop it for years now, android/ndk#1772. Best to just say it works, as we do here, and leave it up to the SDK builders on whether they want to include 32-bit x86, as I understand you do with the Windows toolchain but we do not in the SDK bundle.

As for working with the NDK devs, almost no chance of that, as they will readily tell you [they are understaffed and cannot support outside languages](https://github.com/android/ndk/discussions/1745#discussioncomment-3351867).

> ## Use Cases for Swift on Android

Reviewer: I think that we should re-title this - are we describing the use cases we see or the use cases that we should support?

Reviewer: I do think that there is a difference. These are not the only use cases, its the ones that are currently a focus. I think that this will change over time, and therefore I think that we should re-title this.

Reviewer 2: Ok, please suggest a more appropriate title then, as the ones listed are simply ones we've seen.

The potential applications of Swift on Android are diverse and extend beyond simply sharing code between iOS and Android:

*	Shared Business Logic - A prominent use case is sharing core business logic, algorithms, and data models across iOS and Android applications. This minimizes duplication and ensures consistency in application behavior.
*	Performance-Critical Modules - Swift's strong performance characteristics make it ideal for developing performance-sensitive components, such as image/audio processing, game engines, or other computationally-intensive tasks within an Android app, similar to using C or C++ via the NDK.
*	Cross-Platform Libraries and SDKs - Developers can leverage Swift to build libraries and SDKs that can be easily integrated into existing Android and iOS applications, providing a consistent API surface across platforms.

## Build tool integration

> Integration with existing Android build systems and developer tools is essential in order to provide a usable experience for developers. Dependency management between Swift packages and Android-native dependencies within the existing Gradle build flow will be important. We are looking into providing a Gradle plugin that simplifies the process.

Reviewer: I'm not sure I understand the response. You start by saying no, but then give exactly what I was looking for. It is perfectly fine to say that we intend to tie into Gradle, but what does that mean? Does it mean that we would have a set of Gradle rules, would it be some wrapper, would it be support for SPM to call into Gradle, etc. I would say that we actually don't want to say _how_ it would work (aka the details of the SPM plugin model), that is what the evolution proposal would be aligning up.

> Improved debugging capabilities for Swift libraries on Android are a key focus. This involves enhancing integration with existing debugging tools and protocols. The aim is to make debugging Swift on Android work as well as debugging Swift on other officially supported platforms, including support for Swift-specific metadata and expression evaluation.

Reviewer: The fact that debugging is a focus area is good, but not really a vision. Can we be more explicit about what the integration with existing tools and protocols looks like? I think that being more concrete about the desired state is important. What exactly does debugging Swift specific metadata mean?

Reviewer 2: OK, we can be more specific on that, point taken.

## Testing

Comprehensive testing is important for maintaining code quality on any platform, including running test code on Android emulators and devices. The standard Swift unit and integration testing frameworks, such as XCTest and Swift Testing, work well now, but we will iron out any remaining incompatibilities and establish more Continuous Integration (CI) jobs within the Swift project, such as including Android testing as part of pull request checks, to guarantee ongoing compatibility and stability.

## Bridging to Java and Kotlin

Interoperability between Swift and Android's primary languages- Java and Kotlin- is vital for integrating Swift into existing Android projects and accessing the rich Android SDK. Since JNI is the fundamental mechanism for communicating between native languages like Swift or C/C++ and Java/Kotlin, [automated bridging tools like swift-java](https://github.com/swiftlang/swift-java) should simplify the use of JNI, abstracting away much of its complexity.

> Given Kotlin's prominence on Android, specific bridging mechanisms that handle Kotlin-exclusive features (e.g., suspending functions, specific Jetpack Compose integrations) would be highly beneficial. Recommendations for best practices in designing Swift APIs for optimal interoperability with Kotlin will be provided.

Reviewer: The only change I would recommend is that rather than "best practices ... will be provided" should be more "Documentation covering best practices for ... need to be provided to allow users to easily integrate multiple languages.".

## NDK version support

> Alignment with Android NDK versions is important for stability and access to platform features. We clearly define the Android NDK versions that the latest Swift release/snapshots on Android are compatible with, prioritizing support for Long Term Support (LTS) releases to provide a stable development environment.

Reviewer: I think that we should extend this further to say that we want to have a consistent NDK requirements across host platforms.

## UI recommendations

There will not be a single official UI framework, instead we will acknowledge and support various approaches: the UI for Swift on Android applications will primarily rely on Android's native UI frameworks or existing cross-platform solutions. This means developers will likely choose from:

*	Jetpack Compose - Android's modern, declarative UI toolkit in Kotlin: Swift applications could interact with Compose through robust bridging layers.
*	Android Views (XML-based) - The traditional imperative UI
*	Third-Party Cross-Platform UI Frameworks - Such as Flutter, with Swift as the business logic language via [FlutterSwift](https://github.com/PADL/FlutterSwift), or potentially others that offer C/C++ interop, which Swift can pull in.

> *	Bridging Solutions - Projects such as [Skip](https://github.com/skiptools) provide a declarative SwiftUI-like API surface atop Android's Jetpack Compose, offering a path to native shared cross-platform UI.

Reviewer: Sure, but we have shied away from referring to OSS projects previously as well.

The goal is to ensure Swift works well with these existing UI approaches, allowing developers to select the most suitable option for their project while deploying Swift for the underlying logic.

## Non-goals

Defining what is not a goal is as important as defining what is:

> *	Full SwiftUI/UIKit Port - To reiterate, a direct, full port of existing UI frameworks to Android will not be attempted. The official Swift on Android effort will focus on language and core library support and integrating with existing UI frameworks, not building our own cross-platform UI framework.

Reviewer: I think that this is the wrong way to approach this. While we don't believe that a UIKit replacement needs to be built, the community is free to build something like that.

Reviewer 2: Of course, that is specifically mentioned above. This is specifically saying

1. the workgroup will not work on a GUI toolkit
2. No one GUI toolkit will ever be made the "official recommendation"

> *	Compilation to JVM Bytecode - Direct compilation of Swift to JVM bytecode is not being contemplated. Swift's native AOT compilation model targets machine code, and employing the NDK allows for direct interaction with Android's native layer.

Reviewer 2: In general, people assume that new Android languages will compile to JVM bytecode, like Kotlin does and we explained above, "Most apps are written in Java or Kotlin and run on the Android RunTime (ART), distributed as bytecode and either compiled to machine code when installed or Just-In-Time (JIT) when run."

This bullet point is emphasizing that in addition to what we wrote above, ie "C and C++ code are compiled Ahead-Of-Time (AOT) and packaged as native shared libraries- that's how Swift on Android is also distributed," we will _not_ be pursuing the bytecode approach also, which eg Max suggested on the forums years ago.

> *	Compilation to JVM Bytecode - Direct compilation of Swift to JVM bytecode is not being contemplated. Swift's native AOT compilation model targets machine code, and employing the NDK allows for direct interaction with Android's native layer.

Reviewer: Interesting. I didn't know that had been proposed. Perhaps it is better to extract that into a foreword or separate section that explains how Android works and Swift interacts with that environment? More specifically, I think that what is being suggested is: the vision is to *not* change the fundamentals of Swift, which is a by design an AoT language and we do not intend to change that to a VM based language.

If that is the intent of this, I think that this is something again to be very clear on, because it is very critical that users understand this. More than that, I don't think that it should be a side note in the negative form. That should be pretty clearly stated upfront.

To be clear, I am not making a judgement on the direction, rather I'm just trying to ensure that the vision is clearly shared.

> *	Compilation to JVM Bytecode - Direct compilation of Swift to JVM bytecode is not being contemplated. Swift's native AOT compilation model targets machine code, and employing the NDK allows for direct interaction with Android's native layer.

Reviewer 2: We do clearly state that Swift will be built and distributed like C/C++ in the section above, but added this note at the end to shut the door on the workgroup trying the bytecode approach too.

I don't think this is so important that it needs to be called out in a "separate section," as most app devs don't much care about AoT vs bytecode. For general understanding of readers, we added the explainer above, and for the small minority who may want or expect bytecode too, we added this note.

> *	Compilation to JVM Bytecode - Direct compilation of Swift to JVM bytecode is not being contemplated. Swift's native AOT compilation model targets machine code, and employing the NDK allows for direct interaction with Android's native layer.

Reviewer: I think that we are in agreement then. I don't think that it needs a separate section - it just needs to be clearly stated as a starting point.
